Chiheb feature create silver table postInfo

ChihebDean opened the pull request:
This pull request resolve issue #6

ChihebDean committed:
"fix data model for bronze data table, replace read/write with readStream/writeStream"

ChihebDean committed:
"create silver table postsInfo"

ChihebDean committed:
"fix data model for bronze data table, Fix case class"

ChihebDean committed:
"create delta table postInfo"

ChihebDean committed:
"fix data models for bronze data table and add bronze package under models"

ChihebDean committed:
"create delta table postInfo, fix the test GetPostsInfoSpec, and re-organize packages"

saifeddine1992 commented on src/main/scala/com/fluidcode/configuration/Configuration.scala:
```diff
@@ -63,6 +66,14 @@ case class Configuration(
+    import spark.implicits._
+    val location = s"$rootPath/$database/$postInfoTable"
+    val tableProperties = TableProperties(database, postInfoTable, location)
+    val emptyConf: Seq[Silver] = Seq()
```
naming the case class Silver is too broad. we have multiple tables in the silver layer. Pick a specific name for the case class.

saifeddine1992 commented on src/main/scala/com/fluidcode/configuration/Configuration.scala:
```diff
@@ -73,6 +84,7 @@ object Configuration {
   val DATABASE = "watcher_db"
   val CHECKPOINT_DIR = "checkpoint_dir"
   val BRONZE_TABLE = "RawData"
+  val SILVER_POST_INFO_TABLE = "SilverTable"
```
naming the table "SilverTable" is not accurate. give the table a specific name.

saifeddine1992 commented on src/main/scala/com/fluidcode/models/silver/Silver.scala:
```diff
@@ -0,0 +1,5 @@
+package com.fluidcode.models.silver
+
+case class Silver(comments_disabled: Boolean, edge_media_preview_like_count: Long, edge_media_to_caption_edges_node_text: String,
```
"Silver" is not a specific name for the case class

saifeddine1992 commented on src/main/scala/com/fluidcode/processing/silver/GetPostsInfo.scala:
```diff
@@ -0,0 +1,43 @@
+      col("username"),
+      explode(col("edge_media_to_caption_edges.node.text")).alias("edge_media_to_caption_edges_text")
+    )
+    postsInfo.write.format("delta").mode("overwrite")
+      .option("path", s"${conf.rootPath}.${conf.database}.${conf.postInfoTable}")
+      .saveAsTable(s"${conf.postInfoTable}")
+  }
```
this should be done on the main function, not here.

saifeddine1992 commented on src/main/scala/com/fluidcode/processing/silver/GetPostsInfo.scala:
```diff
@@ -0,0 +1,43 @@
+import org.apache.spark.sql.DataFrame
+import org.apache.spark.sql.functions.{col, explode}
+
+class GetPostsInfo(bronzeData: DataFrame, conf: Configuration) {
```
why use a class instead of an object ?

saifeddine1992 commented on src/main/scala/com/fluidcode/models/Bronze.scala:
```diff
@@ -1,5 +1,24 @@
-
-case class Data(graphImages: Array[Int], GraphProfileInfo: Int)
\ No newline at end of file
+case class Data(graphImages: Array[GraphImagesElements], GraphProfileInfo: ProfileInfo)
+  case class GraphImagesElements(__typename: String, comments: CommentsData, comments_disabled: Boolean, dimensions: DimensionsData,
+                                 display_url: String, edge_media_preview_like: Likes, edge_media_to_caption: Captions,
+                                 edge_media_to_comment: Comments, gating_info: String, id: String, is_video: Boolean,
+                                 location: String, media_preview: String, owner: Owner, shortcode: String, tags: Array[String],
+                                 taken_at_timestamp: Long, thumbnail_resources: Array[ThumbnailElements], thumbnail_src: String,
+                                 urls: Array[String], username: String)
+    case class CommentsData(data: Array[DataElements])
+      case class DataElements(created_at: Long, id: String, owner: OwnerData, text: String)
+        case class OwnerData(id: String, profile_pic_url: String, username: String)
+    case class DimensionsData(height: Long, width: Long)
+    case class Likes(count: Long)
+    case class Captions(edges: Array[EdgesElements])
+      case class EdgesElements(node: NodeData)
+        case class NodeData(text: String)
+    case class Comments(count: Long)
+    case class Owner(id: String)
+    case class ThumbnailElements(config_height: Long, config_width: Long, src: String)
+  case class ProfileInfo(created_time: Long, info: Info, username: String)
+    case class Info(biography: String, followers_count: Long, following_count: Long, full_name: String, id: String,
+                    is_business_account: Boolean, is_joined_recently: Boolean, is_private: Boolean, posts_count: Long,
+                    profile_pic_url: String)
```
why is this not in the Bronze package?

saifeddine1992 commented on src/main/scala/com/fluidcode/models/Gold.scala:
what is this empty object ?

saifeddine1992 commented on src/main/scala/com/fluidcode/processing/bronze/BronzeLayer.scala:
```diff
@@ -7,12 +7,13 @@ class BronzeLayer(conf: Configuration, spark: SparkSession, path: String) {
     bronzeData
-      .write
+      .writeStream
       .format("delta")
```
explore the "checkpoint location" option

ChihebDean committed:
"BronzeLayer and BronzeLayerSpec: read stream using schema, use writeStream to write bronze data, and add option checkpointLocation "

ChihebDean committed:
"change database, and bronze table names"

ChihebDean committed:
"rename case class Silver to PostsData, edit silver functions, and tests"

saifeddine1992 commented on src/main/scala/com/fluidcode/processing/bronze/BronzeLayer.scala:
```diff
@@ -1,19 +1,21 @@
       .outputMode("append")
-      .option("path", s"${conf.rootPath}/${conf.database}/${conf.bronzeTable}")
-      .start()
+      .option("checkpointLocation", "checkpoint_dir")
```
not quite right. If you keep it like this, other checkpoints from other tables will be written under the same directory.
we need a checkpoint location for each table. 
Also, the checkpoint directory must be extracted from Configuration and not hardcoded

saifeddine1992 commented on src/test/scala/com/fluidcode/processing/silver/GetPostsInfoSpec.scala:
```diff
@@ -4,29 +4,34 @@ import com.fluidcode.configuration.Configuration
   with DeltaExtendedSparkSession  {
 
-  test("getPostsInfo should create  PostsInfoTable from Bronze Layer" ) {
+  test("getPostsInfo should select post info data from Bronze Layer" ) {
```
this should be done in the other test class, not here. 
for each object or class, we need to have a mirroring test class, not 2.

saifeddine1992 commented on src/main/scala/com/fluidcode/models/golden/Gold.scala:
```diff
@@ -0,0 +1,5 @@
+package com.fluidcode.models.golden
```
gold

saifeddine1992 commented on src/main/scala/com/fluidcode/processing/silver/Silver.scala:
```diff
@@ -0,0 +1,24 @@
+    conf.init(spark)
+
+    val bronzeTablePath = s"${conf.rootPath}/${conf.database}/${conf.bronzeTable}/"
+    val bronzeData = spark.read.format("delta").load(bronzeTablePath)
```
stream

saifeddine1992 commented on src/test/scala/com/fluidcode/processing/bronze/BronzeLayerSpec.scala:
```diff
@@ -1,27 +1,31 @@
-      
-      val path = "phil.coutinho-1-test.json"
+
+      val path = "c:/tmp/StreamSource/bronze"
```
doesn't align with a production env

saifeddine1992 commented on src/main/scala/com/fluidcode/processing/silver/GetPostsInfoAndCreateTable.scala:
```diff
@@ -0,0 +1,47 @@
+    postsInfo
+  }
+  def createPostsInfoTable(postsInfo: DataFrame, conf: Configuration): Unit = {
+    postsInfo.write.format("delta").mode("overwrite")
```
replace write with writeStream
add readStream

ChihebDean committed:
"BronzeLayer: change checkpoint directory.
BronzeLayerSpec: create a sample df for testing, then read raw data from sample path"

ChihebDean committed:
"read bronze data inside function createPostsInfoTable, and change read to write to writeStream"

saifeddine1992 commented on src/main/scala/com/fluidcode/processing/bronze/BronzeLayer.scala:
```diff
@@ -10,12 +10,10 @@ class BronzeLayer(conf: Configuration, spark: SparkSession, path: String) {
-    bronzeData
-      .writeStream
+    bronzeData.writeStream
+      .option("checkpointLocation", s"${conf.checkpointDir}")
```
the checkpoint location is no longer hard coded, but we stil have the same problem. all checkpoints will be written in the same place.

saifeddine1992 commented on src/main/scala/com/fluidcode/processing/silver/GetPostsInfoAndCreateTable.scala:
```diff
@@ -38,10 +39,16 @@ object GetPostsInfoAndCreateTable {
+    val postsInfo = getPostsInfo(bronzeData)
+
+    postsInfo.writeStream
+      .option("checkpointLocation", s"${conf.checkpointDir}")
```
checkpoints of the bronze table will be written here too, this should change to a checkpoint location specific to the table.

saifeddine1992 commented on src/main/scala/com/fluidcode/processing/silver/Silver.scala:
```diff
@@ -12,13 +12,8 @@ object Silver {
-    createPostsInfoTable(postsInfoData, conf)
-
 
+    val bronzeDataPath = args(1)
```
you can create the path of the bronze data from Configuration.

saifeddine1992 commented on src/test/scala/com/fluidcode/processing/bronze/BronzeLayerSpec.scala:
```diff
@@ -16,7 +17,21 @@ class BronzeLayerSpec extends QueryTest
-      val path = "c:/tmp/StreamSource/bronze"
+      import sparkSession.implicits._
+
+      val path = s"${conf.rootPath}/${conf.database}/${conf.sample}"
```
this path is created for the sake of testing, it shouldn't be added to our Configuration.

saifeddine1992 commented on src/test/scala/com/fluidcode/processing/silver/GetPostsInfoAndCreateTableSpec.scala:
```diff
@@ -9,25 +12,83 @@ import org.apache.spark.sql.test.SharedSparkSession
-      val bronzeData = spark.read.format("delta").load(s"${conf.rootPath}/${conf.database}/${conf.bronzeTable}")
-      val postsInfo = getPostsInfo(bronzeData)
-      createPostsInfoTable(postsInfo, conf)
+      createPostsInfoTable(s"${conf.rootPath}/${conf.database}/${conf.bronzeTable}", conf, spark)
```
this function can take "conf" and "spark" only, no need to give it a path as an argument as you are able to extract the path from Configuration.

Commit:
"fix the checkpoint location for bronze table and post info table"

ChihebDean committed:
"Edit sample path, and read bronze data path directly from configuration, not as a function's argument"

saifeddine1992 commented on src/main/scala/com/fluidcode/processing/bronze/BronzeLayer.scala:
```diff
@@ -11,7 +12,7 @@ class BronzeLayer(conf: Configuration, spark: SparkSession, path: String) {
 
     bronzeData.writeStream
-      .option("checkpointLocation", s"${conf.checkpointDir}")
+      .option("checkpointLocation", s"${conf.rootPath}/${conf.database}/${conf.bronzeTable}/$CHECKPOINT_DIR")
```
same problem remains. start from the checkpoint directory and then specify the table.

saifeddine1992 commented on src/main/scala/com/fluidcode/processing/silver/GetPostsInfoAndCreateTable.scala:
```diff
@@ -39,13 +40,13 @@ object GetPostsInfoAndCreateTable {
 
     postsInfo.writeStream
-      .option("checkpointLocation", s"${conf.checkpointDir}")
+      .option("checkpointLocation", s"${conf.rootPath}/${conf.database}/${conf.bronzeTable}/$CHECKPOINT_DIR")
```
same problem. nothing should be hardcoded.
PS: start from the checkpoint directory then specify the table.

ChihebDean replied:
$CHECKPOINT_DIR is hardcoded?

saifeddine1992 commented on src/test/scala/com/fluidcode/processing/bronze/BronzeLayerSpec.scala:
```diff
@@ -19,7 +19,7 @@ class BronzeLayerSpec extends QueryTest
       import sparkSession.implicits._
 
-      val path = s"${conf.rootPath}/${conf.database}/${conf.sample}"
+      val path = s"${conf.rootPath}/${conf.database}/sample"
```
The path is still a part of the project configuration in this case.  it needs to be totally separated from the Configuration and only linked to the test.

saifeddine1992 commented on src/test/scala/com/fluidcode/processing/silver/GetPostsInfoAndCreateTableSpec.scala:
```diff
@@ -22,7 +22,7 @@ class GetPostsInfoAndCreateTableSpec extends QueryTest
       import sparkSession.implicits._
 
-      val path = s"${conf.rootPath}/${conf.database}/${conf.sample}"
+      val path = s"${conf.rootPath}/${conf.database}/sample"
```
same.

ChihebDean replied:
Done

saifeddine1992 commented:
The test of the function GetPostData is wrong. the input is a DataFrame and the output is also a DataFrame. you don't need to write intermediate data in the sample directory. just create an input DataFrame and an expectedResult DataFrame and do an ordinary unit test.

ChihebDean committed:
"fix checkpoint location for bronze table and post info table"

ChihebDean committed:
"fix sample path and getPostsInfo function test"

saifeddine1992 commented on src/test/scala/com/fluidcode/processing/silver/GetPostsInfoAndCreateTableSpec.scala:
```diff
@@ -0,0 +1,82 @@
+      assert(result.except(expectedResult).isEmpty)
+  }
+
+  test("getPostsInfo should create post info table from Bronze Layer" ) {
```
test description is wrong

ChihebDean committed:
Revert ""fix data model for bronze data table, Fix case class""

This reverts commit 3dff60e1

ChihebDean committed:
"delete"

ChihebDean committed:
"fix checkpoint location path for bronze/posts info table"

ChihebDean committed:
"change function name in test from getPostsInfo to createPostsInfoTable"

saifeddine1992 commented on src/test/scala/com/fluidcode/processing/silver/GetPostsInfoAndCreateTableSpec.scala:
```diff
@@ -0,0 +1,82 @@
+package com.fluidcode.processing.silver
+
+import com.fluidcode.configuration.Configuration
+import com.fluidcode.models.bronze.{Captions, Comments, CommentsData, Data, DataElements, DimensionsData, EdgesElements,
+                            GraphImagesElements, Info, Likes, NodeData, Owner, OwnerData, ProfileInfo, ThumbnailElements}
```
try with "._" instead if naming all case classes in the import

ChihebDean replied:
Ok!

ChihebDean committed:
"import all case classes using ._"

ChihebDean committed:
"rename silver model case class, and silver Object GetPostsInfoAndCreateTable"

saifeddine1992 commented:
pull request has to have the same name of the issue or a close name.
